pkg/controllers/deprovisioning/controller.go

```diff
@@ -0,0 +1,513 @@
+	}
+
+	metrics.NodesCreatedCounter.WithLabelValues(metrics.DeprovisioningReason).Inc()
+
+	// We have the new node created at the API server so mark the old node for deletion
+	c.cluster.MarkForDeletion(oldNode.Name)
+
+	var k8Node v1.Node
+	// Wait for the node to be ready
+	var once sync.Once
+	if err := retry.Do(func() error {
+		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodeNames[0]}, &k8Node); err != nil {
+			return fmt.Errorf("getting node, %w", err)
+		}
+		once.Do(func() {
+			c.recorder.Publish(LaunchingNode(&k8Node, action.String()))
+		})
+
+		if _, ok := k8Node.Labels[v1alpha5.LabelNodeInitialized]; !ok {
+			// make the user aware of why deprovisioning is paused
+			c.recorder.Publish(WaitingOnReadiness(&k8Node))
+			return fmt.Errorf("node is not initialized")
+		}
+		return nil
+	}, waitRetryOptions...); err != nil {
+		// node never become ready, so uncordon the node we were trying to delete and report the error
+		c.cluster.UnmarkForDeletion(oldNode.Name)
+		return multierr.Combine(c.setNodeUnschedulable(ctx, oldNode.Name, false),
+			fmt.Errorf("timed out checking node readiness, %w", err))
+	}
+	return nil
+}
+
+// calculateLifetimeRemaining calculates the fraction of node lifetime remaining in the range [0.0, 1.0].  If the TTLSecondsUntilExpired
+// is non-zero, we use it to scale down the disruption costs of nodes that are going to expire.  Just after creation, the
+// disruption cost is highest and it approaches zero as the node ages towards its expiration time.
+func (c *Controller) calculateLifetimeRemaining(node candidateNode) float64 {
+	remaining := 1.0
+	if node.provisioner.Spec.TTLSecondsUntilExpired != nil {
+		ageInSeconds := c.clock.Since(node.CreationTimestamp.Time).Seconds()
+		totalLifetimeSeconds := float64(*node.provisioner.Spec.TTLSecondsUntilExpired)
+		lifetimeRemainingSeconds := totalLifetimeSeconds - ageInSeconds
+		remaining = clamp(0.0, lifetimeRemainingSeconds/totalLifetimeSeconds, 1.0)
+	}
+	return remaining
+}
+
+func (c *Controller) setNodeUnschedulable(ctx context.Context, nodeName string, isUnschedulable bool) error {
+	var node v1.Node
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodeName}, &node); err != nil {
+		return fmt.Errorf("getting node, %w", err)
+	}
+
+	// node is being deleted already, so no need to un-cordon
+	if !isUnschedulable && !node.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// already matches the state we want to be in
+	if node.Spec.Unschedulable == isUnschedulable {
+		return nil
+	}
+
+	persisted := node.DeepCopy()
+	node.Spec.Unschedulable = isUnschedulable
+	if err := c.kubeClient.Patch(ctx, &node, client.MergeFrom(persisted)); err != nil {
+		return fmt.Errorf("patching node %s, %w", node.Name, err)
+	}
+	return nil
+}
+
+// mapNodes maps from a list of *v1.Node to candidateNode
+func (c *Controller) mapNodes(nodes []*v1.Node, candidateNodes []candidateNode) []candidateNode {
```

Comment:
helper function?

Reply:
This is just an intersection so why not `sets.Intersection`?

Reply:
I'll move it to the helper folder since it doesn't need to be a member function, but won't implement it as a sets function for the same as above.